File: frontend/lib/dataPlatform.ts
import { api } from "@/lib/api";

export type DataSource = {
  id: number;
  owner_user_id: number;
  name: string;
  kind: "website" | "api";
  base_url: string;
  is_enabled: boolean;
  crawl_config: Record<string, unknown>;
  transform_config: Record<string, unknown>;
  classify_config: Record<string, unknown>;
  retention_config: Record<string, unknown>;
  last_crawl_started_at?: string | null;
  last_crawl_finished_at?: string | null;
  last_crawl_status?: string | null;
  created_at: string;
  updated_at?: string | null;
};

export type CrawlJob = {
  id: number;
  owner_user_id: number;
  source_id: number;
  status: "queued" | "running" | "succeeded" | "failed" | "cancelled";
  message?: string | null;
  consent_received: boolean;
  policy_context: Record<string, unknown>;
  strategy_override: Record<string, unknown>;
  started_at?: string | null;
  finished_at?: string | null;
  discovered_count: number;
  fetched_count: number;
  stored_count: number;
  skipped_count: number;
  failed_count: number;
  created_at: string;
};

export type DocumentItem = {
  id: number;
  owner_user_id: number;
  source_id: number;
  job_id?: number | null;
  url: string;
  canonical_url?: string | null;
  title?: string | null;
  content_text: string;
  extracted: Record<string, unknown>;
  language?: string | null;
  category?: string | null;
  tags: string[];
  quality_score?: number | null;
  quality_flags: string[];
  pii_redacted: boolean;
  first_seen_at: string;
  last_seen_at: string;
};

export async function createSource(params: {
  token: string;
  name: string;
  kind: "website" | "api";
  base_url: string;
  crawl_config?: Record<string, unknown>;
  transform_config?: Record<string, unknown>;
  classify_config?: Record<string, unknown>;
  retention_config?: Record<string, unknown>;
}): Promise<DataSource> {
  return api.post<DataSource>("/api/data-platform/sources", {
    name: params.name,
    kind: params.kind,
    base_url: params.base_url,
    crawl_config: params.crawl_config || {},
    transform_config: params.transform_config || {},
    classify_config: params.classify_config || {},
    retention_config: params.retention_config || {},
  }, { token: params.token });
}

export async function listSources(token: string): Promise<DataSource[]> {
  return api.get<DataSource[]>("/api/data-platform/sources", { token });
}

export async function startJob(params: {
  token: string;
  source_id: number;
  consent: boolean;
  policy_context?: Record<string, unknown>;
  strategy_override?: Record<string, unknown>;
}): Promise<CrawlJob> {
  return api.post<CrawlJob>(`/api/data-platform/sources/${params.source_id}/jobs`, {
    consent: params.consent,
    policy_context: params.policy_context || {},
    strategy_override: params.strategy_override || {},
  }, { token: params.token });
}

export async function listJobs(token: string, source_id?: number): Promise<CrawlJob[]> {
  const params = source_id ? { source_id } : undefined;
  return api.get<CrawlJob[]>("/api/data-platform/jobs", { token, params });
}

export async function searchDocuments(params: {
  token: string;
  q?: string;
  source_id?: number;
  category?: string;
  tags?: string[];
  quality_min?: number;
  limit?: number;
  offset?: number;
}): Promise<{ total: number; items: DocumentItem[] }> {
  const queryParams: Record<string, any> = {
    limit: params.limit ?? 25,
    offset: params.offset ?? 0,
  };
  if (params.q) queryParams.q = params.q;
  if (typeof params.source_id === "number") queryParams.source_id = params.source_id;
  if (params.category) queryParams.category = params.category;
  if (params.tags && params.tags.length) queryParams.tags = params.tags.join(',');
  if (typeof params.quality_min === "number") queryParams.quality_min = params.quality_min;
  
  return api.get<{ total: number; items: DocumentItem[] }>("/api/data-platform/documents", {
    token: params.token,
    params: queryParams
  });
}

export async function qualitySummary(token: string, source_id?: number): Promise<any> {
  const params = source_id ? { source_id } : undefined;
  return api.get<any>("/api/data-platform/quality/summary", { token, params });
}

File: backend/app/schemas/data_platform.py
from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class DataSourceCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    kind: str = Field(pattern="^(website|api)$")
    base_url: str = Field(min_length=1, max_length=1024)

    crawl_config: Dict[str, Any] = Field(default_factory=dict)
    transform_config: Dict[str, Any] = Field(default_factory=dict)
    classify_config: Dict[str, Any] = Field(default_factory=dict)
    retention_config: Dict[str, Any] = Field(default_factory=dict)


class DataSourceUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=120)
    base_url: Optional[str] = Field(default=None, min_length=1, max_length=1024)
    is_enabled: Optional[bool] = None

    crawl_config: Optional[Dict[str, Any]] = None
    transform_config: Optional[Dict[str, Any]] = None
    classify_config: Optional[Dict[str, Any]] = None
    retention_config: Optional[Dict[str, Any]] = None


class DataSourceOut(BaseModel):
    id: int
    owner_user_id: int
    name: str
    kind: str
    base_url: str
    is_enabled: bool

    crawl_config: Dict[str, Any]
    transform_config: Dict[str, Any]
    classify_config: Dict[str, Any]
    retention_config: Dict[str, Any]

    last_crawl_started_at: Optional[datetime] = None
    last_crawl_finished_at: Optional[datetime] = None
    last_crawl_status: Optional[str] = None

    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class CrawlJobStart(BaseModel):
    consent: bool = False
    policy_context: Dict[str, Any] = Field(default_factory=dict)
    strategy_override: Dict[str, Any] = Field(default_factory=dict)


class CrawlJobOut(BaseModel):
    id: int
    owner_user_id: int
    source_id: int
    status: str
    message: Optional[str] = None

    consent_received: bool
    policy_context: Dict[str, Any]
    strategy_override: Dict[str, Any]

    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None

    discovered_count: int
    fetched_count: int
    stored_count: int
    skipped_count: int
    failed_count: int

    created_at: datetime

    class Config:
        from_attributes = True


class DocumentOut(BaseModel):
    id: int
    owner_user_id: int
    source_id: int
    job_id: Optional[int] = None

    url: str
    canonical_url: Optional[str] = None
    title: Optional[str] = None
    content_text: str
    extracted: Dict[str, Any]

    language: Optional[str] = None
    category: Optional[str] = None
    tags: List[str]

    quality_score: Optional[float] = None
    quality_flags: List[str]
    pii_redacted: bool

    first_seen_at: datetime
    last_seen_at: datetime

    class Config:
        from_attributes = True


class DocumentSearchResponse(BaseModel):
    total: int
    items: List[DocumentOut]

